proof/courier.go

```diff
@@ -295,53 +353,294 @@ func (h *HashMailCourier) DeliverProof(ctx context.Context, addr address.Taro,
+	for _, timestamp := range timestamps {
+		if timestamp.After(latestTimestamp) {
+			latestTimestamp = timestamp
+		}
+	}
+
+	return time.Since(latestTimestamp)
+}
+
+// BackoffExecError is an error returned when the backoff execution fails.
+// This error wraps the underlying error returned by the execution function.
+// It allows the porter to determine whether the state machine should be halted
+// or not.
+type BackoffExecError struct {
+	execErr error
+}
+
+func (e *BackoffExecError) Error() string {
+	return fmt.Sprintf("backoff exec error: %s", e.execErr.Error())
+}
+
+// backoffExec attempts to execute the given `exec` function using a repeating
+// backoff time delayed strategy. The backoff strategy is used to ensure
+// that we don't spam the hashmail service with proof delivery attempts.
+func (h *HashMailCourier) backoffExec(ctx context.Context,
+	exec func() error) error {
+
+	var (
+		backoff          = h.cfg.BackoffCfg.InitialBackoff
+		numTries         = h.cfg.BackoffCfg.NumTries
+		maxBackoff       = h.cfg.BackoffCfg.MaxBackoff
+		err        error = nil
+	)
+
+	for i := 0; i < numTries; i++ {
+		// Don't delay at first try. Wait before reattempting
+		// delivery if a backoff duration is specified.
+		if i > 0 && backoff != 0 {
+			// Notify subscribers that backoff wait is about to
+			// commence.
+			transferEvent := NewReceiverProofBackoffWaitEvent(
+				backoff, int64(i),
+			)
+			h.publishSubscriberEvent(transferEvent)
+
+			err = h.wait(backoff)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = exec()
+		if err != nil {
+			err = fmt.Errorf("error executing backoff procedure: "+
+				"%w", &BackoffExecError{execErr: err})
+
+			// Function execution wasn't successful, increase the
+			// time we'll wait until the next try.
+			backoff *= 2
+
+			// Cap the backoff at the maximum backoff.
+			if backoff > maxBackoff {
+				backoff = maxBackoff
+			}
+			log.Debugf("Receiver proof delivery failed with "+
+				"error. Backing off for %s: %v", backoff, err)
+
+			if i < numTries-1 {
+				log.Infof("Receiver proof delivery failed, "+
+					"will try again after backoff delay: "+
+					"%v", backoff)
+			}
+
+			// Ready to retry proof delivery if applicable.
+			continue
+		}
+
+		// If the proof delivery was successful, we can exit the loop.
+		break
+	}
+	if err != nil {
+		return fmt.Errorf("receiver proof delivery failed; count "+
+			"retries attempted: %d; %w", numTries, err)
+	}
+
+	return nil
+}
+
+// publishSubscriberEvent publishes an event to all subscribers.
+func (h *HashMailCourier) publishSubscriberEvent(event chanutils.Event) {
+	// Lock the subscriber mutex to ensure that we don't modify the
+	// subscriber map while we're iterating over it.
+	h.subscriberMtx.Lock()
+	defer h.subscriberMtx.Unlock()
+
+	for _, sub := range h.subscribers {
+		sub.NewItemCreated.ChanIn() <- event
+	}
+}
+
+// wait blocks for a given amount of time.
+//
+// TODO(ffranr): add HashMailCourier quit signal?
```

Comment:
We could delete this function and simply use `<-time.After(backoff)` wherever `h.wait(backoff)` is used. However, I think we should always have at least two or entries in the select, on of them being the time.After, ctx.Done and the s.quit

Reply:
Yes, we definitely need a quit here.

Reply:
`HashMailCourier` doesn't have a quit channel. The `ChainPorter` does have a quit channel, but I don't think anything writes to it.

In adding a quit channel to `HashMailCourier`, is the idea to write to that channel if `ChainPorter.stop()` is called? Do I need to make sure that each time the `ChainPorter`'s quit channel is read, it leads to calling `ChainPorter.stop()` and thus `HashMailCourier`'s quit channel is fired?

(In the latest commits I've added context handling only.)